src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,38 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        private static bool TryEnsureSufficientExecutionStack<TArg>(int recursionDepth, Func<TArg, bool> throwOnFailure, TArg arg)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure(arg))
+            {
+                return false;
+            }
+        }
+
+        internal static TResult Execute<TResult, TArg1, TArg2>(ref int recursionDepth, Func<TArg1, bool> throwOnFailure, Func<TArg1, TArg2, TResult> execute, TArg1 arg1, TArg2 arg2)
+        {
+            recursionDepth++;
+            TResult result;
+            if (TryEnsureSufficientExecutionStack(recursionDepth, throwOnFailure, arg1))
+            {
+                result = execute(arg1, arg2);
+            }
+            else
+            {
+                var task = Task.Run(() => execute(arg1, arg2));
```

Comment:
You might want to change this to:
```C#
TArg1 localArg1 = arg1;
TArg2 localArg2 = arg2;
var task = Task.Run(() => execute(localArg1, localArg2));
```
Otherwise, calling this `Execute<>` method is always going to allocate a closure to store `arg1`/`arg2` even though you should only rarely enter the `else`.

Reply:
Good catch, thanks. (It looks like `Execute` currently uses 48 bytes on x64 and I need to reduce that overhead.)

---
In reply to: [357836468](https://github.com/dotnet/roslyn/pull/40367#discussion_r357836468) [](ancestors = 357836468)

Reply:
static-lambdas will help here :)  (hint hint, nudge nudge)